## 1. Data overview <a id='data_review'></a>

Open the data and explore it.

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('vehicles_us.csv')
data 

price  model_year           model  condition  cylinders fuel  odometer  \
0       9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1      25500         NaN      ford f-150       good        6.0  gas   88705.0   
2       5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3       1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4      14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
...      ...         ...             ...        ...        ...  ...       ...   
51520   9249      2013.0   nissan maxima   like new        6.0  gas   88136.0   
51521   2700      2002.0     honda civic    salvage        4.0  gas  181500.0   
51522   3950      2009.0  hyundai sonata  excellent        4.0  gas  128000.0   
51523   7455      2013.0  toyota corolla       good        4.0  gas  139573.0   
51524   6300      2014.0   nissan altima       good        4.0  gas       NaN   

      transmission    type paint_color  is_4wd date_posted  days_listed  
0        automatic     SUV         NaN     1.0  2018-06-23           19  
1        automatic  pickup       white     1.0  2018-10-19           50  
2        automatic   sedan         red     NaN  2019-02-07           79  
3        automatic  pickup         NaN     NaN  2019-03-22            9  
4        automatic   sedan       black     NaN  2019-04-02           28  
...            ...     ...         ...     ...         ...          ...  
51520    automatic   sedan       black     NaN  2018-10-03           37  
51521    automatic   sedan       white     NaN  2018-11-14           22  
51522    automatic   sedan        blue     NaN  2018-11-15           32  
51523    automatic   sedan       black     NaN  2018-07-02           71  
51524    automatic   sedan         NaN     NaN  2018-06-05           10  

[51525 rows x 13 columns]

In [3]:
data.head(10)

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
5  14990      2014.0    chrysler 300  excellent        6.0  gas   57954.0   
6  12990      2015.0    toyota camry  excellent        4.0  gas   79212.0   
7  15990      2013.0     honda pilot  excellent        6.0  gas  109473.0   
8  11500      2012.0     kia sorento  excellent        4.0  gas  104174.0   
9   9200      2008.0     honda pilot  excellent        NaN  gas  147191.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28  
5    automatic   sedan       black     1.0  2018-06-20           15  
6    automatic   sedan       white     NaN  2018-12-27           73  
7    automatic     SUV       black     1.0  2019-01-07           68  
8    automatic     SUV         NaN     1.0  2018-07-16           19  
9    automatic     SUV        blue     1.0  2019-02-15           17

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [5]:
data.columns

Index(['price', 'model_year', 'model', 'condition', 'cylinders', 'fuel',
       'odometer', 'transmission', 'type', 'paint_color', 'is_4wd',
       'date_posted', 'days_listed'],
      dtype='object')

The table contains 13 columns. They store 3 types of data: float, int, object.

According to the documentation:

- 'price' 
- 'model_year' 
- 'model' 
- 'condition'
- 'cylinders' 
- 'fuel' 
- 'odometer' 
- 'transmission'
- 'type' 
- 'paint_color' 
- 'is_4wd'
- 'date_posted'
- 'days_listed'

## 2. Missing values <a id='intro'></a>
Lets, find the number of missing values in the table.

In [6]:
data.isna().sum()

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

We have missing values in columns:  
- model_year - 3619
- cylinders - 5260 
- odometer - 7892      
- paint_color - 9267
- is_4wd - 25953

Not all missing values affect the research. For instance, the missing values in odometer is not critical. We can simply replace them with clear markers.

But another missing values can affect the comparison of car preferences. 

In [7]:
data.model_year.unique()

array([2011.,   nan, 2013., 2003., 2017., 2014., 2015., 2012., 2008.,
       2018., 2009., 2010., 2007., 2004., 2005., 2001., 2006., 1966.,
       1994., 2019., 2000., 2016., 1993., 1999., 1997., 2002., 1981.,
       1995., 1996., 1975., 1998., 1985., 1977., 1987., 1974., 1990.,
       1992., 1991., 1972., 1967., 1988., 1969., 1989., 1978., 1965.,
       1979., 1968., 1986., 1980., 1964., 1963., 1984., 1982., 1973.,
       1970., 1955., 1971., 1976., 1983., 1954., 1962., 1948., 1960.,
       1908., 1961., 1936., 1949., 1958., 1929.])

In [8]:
sorted(data.cylinders.unique())

[3.0, 4.0, 5.0, 6.0, 8.0, nan, 10.0, 12.0]

In [9]:
sorted(data.odometer.unique())

[0.0,
 1.0,
 2.0,
 3.0,
 5.0,
 6.0,
 7.0,
 10.0,
 13.0,
 14.0,
 20.0,
 23.0,
 25.0,
 27.0,
 28.0,
 35.0,
 42.0,
 43.0,
 50.0,
 54.0,
 56.0,
 62.0,
 63.0,
 65.0,
 66.0,
 68.0,
 73.0,
 76.0,
 77.0,
 80.0,
 86.0,
 87.0,
 88.0,
 91.0,
 98.0,
 100.0,
 101.0,
 102.0,
 103.0,
 104.0,
 107.0,
 114.0,
 116.0,
 117.0,
 118.0,
 120.0,
 121.0,
 124.0,
 126.0,
 128.0,
 129.0,
 130.0,
 132.0,
 136.0,
 137.0,
 138.0,
 140.0,
 141.0,
 145.0,
 147.0,
 153.0,
 155.0,
 156.0,
 159.0,
 160.0,
 161.0,
 162.0,
 164.0,
 166.0,
 167.0,
 168.0,
 172.0,
 174.0,
 175.0,
 176.0,
 185.0,
 192.0,
 195.0,
 198.0,
 200.0,
 205.0,
 206.0,
 207.0,
 210.0,
 214.0,
 215.0,
 220.0,
 223.0,
 229.0,
 236.0,
 237.0,
 254.0,
 255.0,
 270.0,
 290.0,
 345.0,
 387.0,
 432.0,
 500.0,
 689.0,
 711.0,
 760.0,
 950.0,
 977.0,
 998.0,
 999.0,
 1003.0,
 1142.0,
 1181.0,
 1277.0,
 1300.0,
 1399.0,
 1440.0,
 1566.0,
 1574.0,
 1756.0,
 2000.0,
 2004.0,
 2045.0,
 2050.0,
 2073.0,
 2118.0,
 2196.0,
 2470.0,
 2477.0,
 2500.0,
 2534.0,
 2550

In [10]:
data.paint_color.unique()

array([nan, 'white', 'red', 'black', 'blue', 'grey', 'silver', 'custom',
       'orange', 'yellow', 'brown', 'green', 'purple'], dtype=object)

In [11]:
data.is_4wd.unique()

array([ 1., nan])

In [12]:
columns_to_replace = ['paint_color']

for column in columns_to_replace:
    print(column)
    data[column] = data[column].fillna('unknown')
    print('Missing values in ', column, 'are replaced') 

paint_color
Missing values in  paint_color are replaced


In [13]:
data.paint_color.unique()

array(['unknown', 'white', 'red', 'black', 'blue', 'grey', 'silver',
       'custom', 'orange', 'yellow', 'brown', 'green', 'purple'],
      dtype=object)

In [14]:
column_to_replace = ['is_4wd']

for column in column_to_replace:
    print(column)
    data[column] = data[column].fillna('0')
    print('Missing values in ', column, 'are replaced') 

is_4wd
Missing values in  is_4wd are replaced


In [15]:
data.is_4wd.unique()

array([1.0, '0'], dtype=object)

In [16]:
data['model_year'] = data['model_year'].fillna(data.groupby(['model'])['model_year'].transform('median'))
print(data['model_year'])

0        2011.0
1        2011.0
2        2013.0
3        2003.0
4        2017.0
          ...  
51520    2013.0
51521    2002.0
51522    2009.0
51523    2013.0
51524    2014.0
Name: model_year, Length: 51525, dtype: float64


In [17]:
data.model_year.unique()

array([2011. , 2013. , 2003. , 2017. , 2014. , 2015. , 2012. , 2008. ,
       2018. , 2009. , 2010. , 2007. , 2004. , 2005. , 2001. , 2006. ,
       1966. , 1994. , 2019. , 2000. , 2016. , 1993. , 1999. , 2006.5,
       1997. , 2002. , 1981. , 1995. , 1996. , 1975. , 1998. , 1985. ,
       1977. , 1987. , 1974. , 1990. , 1992. , 1991. , 1972. , 1967. ,
       1988. , 1969. , 1989. , 1978. , 1965. , 1979. , 1968. , 1986. ,
       1980. , 1964. , 1963. , 1984. , 1982. , 2010.5, 1973. , 1970. ,
       1955. , 1971. , 1976. , 1983. , 1954. , 1962. , 1948. , 1960. ,
       1908. , 1961. , 1936. , 1949. , 1958. , 1929. ])

In [18]:
data['odometer'] = data['odometer'].fillna(data.groupby(['model'])['odometer'].transform('median'))
print(data['odometer'])

0        145000.0
1         88705.0
2        110000.0
3        121928.0
4         80903.0
           ...   
51520     88136.0
51521    181500.0
51522    128000.0
51523    139573.0
51524    100319.5
Name: odometer, Length: 51525, dtype: float64


In [19]:
data['odometer'] = data['odometer'].fillna(data.groupby(['model_year'])['odometer'].transform('median'))
print(data['odometer'])

0        145000.0
1         88705.0
2        110000.0
3        121928.0
4         80903.0
           ...   
51520     88136.0
51521    181500.0
51522    128000.0
51523    139573.0
51524    100319.5
Name: odometer, Length: 51525, dtype: float64


In [20]:
data['cylinders'] = data['cylinders'].fillna(data.groupby(['model'])['cylinders'].transform('median'))
print(data['cylinders'])

0        6.0
1        6.0
2        4.0
3        8.0
4        4.0
        ... 
51520    6.0
51521    4.0
51522    4.0
51523    4.0
51524    4.0
Name: cylinders, Length: 51525, dtype: float64


Let's make sure the table contains no more missing values.

In [21]:
data.isna().sum()

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

### 3. Duplicates <a id='duplicates'></a>
Find the number of obvious duplicates in the table:

In [23]:
data.duplicated().sum()

0

We have no obvious duplicates

Let's get rid of implicit duplicates in the "model" column.

In [24]:
sorted(data.model.unique())

['acura tl',
 'bmw x5',
 'buick enclave',
 'cadillac escalade',
 'chevrolet camaro',
 'chevrolet camaro lt coupe 2d',
 'chevrolet colorado',
 'chevrolet corvette',
 'chevrolet cruze',
 'chevrolet equinox',
 'chevrolet impala',
 'chevrolet malibu',
 'chevrolet silverado',
 'chevrolet silverado 1500',
 'chevrolet silverado 1500 crew',
 'chevrolet silverado 2500hd',
 'chevrolet silverado 3500hd',
 'chevrolet suburban',
 'chevrolet tahoe',
 'chevrolet trailblazer',
 'chevrolet traverse',
 'chrysler 200',
 'chrysler 300',
 'chrysler town & country',
 'dodge charger',
 'dodge dakota',
 'dodge grand caravan',
 'ford econoline',
 'ford edge',
 'ford escape',
 'ford expedition',
 'ford explorer',
 'ford f-150',
 'ford f-250',
 'ford f-250 sd',
 'ford f-250 super duty',
 'ford f-350 sd',
 'ford f150',
 'ford f150 supercrew cab xlt',
 'ford f250',
 'ford f250 super duty',
 'ford f350',
 'ford f350 super duty',
 'ford focus',
 'ford focus se',
 'ford fusion',
 'ford fusion se',
 'ford mustang',
 '

In [25]:
for car_model in data.model.unique():
    if 'ford f-' in car_model or 'ford f' in car_model:
        print(car_model)

ford f-150
ford fusion se
ford focus
ford f150 supercrew cab xlt
ford f-250 sd
ford f250 super duty
ford f-350 sd
ford f-250
ford f150
ford f350 super duty
ford fusion
ford f-250 super duty
ford focus se
ford f250
ford f350


In [26]:
def replace_wrong_models(wrong_models, correct_models):
    for element in wrong_models:
        data['model'] = data['model'].replace(element, correct_models)    

In [27]:
replace_wrong_models(['ford f-250 sd', 'ford f250 super duty'], 'ford f-250 super duty')
replace_wrong_models(['ford f150', 'fford f-150'], 'ford f-150')
replace_wrong_models(['ford f250'], 'ford f-250')
replace_wrong_models(['ford f350 super duty', 'ford f-350 sd'], 'ford f-350 super duty')

In [28]:
for car_model in data.model.unique():
    if 'ford f-' in car_model or 'ford f' in car_model:
        print(car_model)

ford f-150
ford fusion se
ford focus
ford f150 supercrew cab xlt
ford f-250 super duty
ford f-350 super duty
ford f-250
ford fusion
ford focus se
ford f350


Let's sure the duplicate names were removed. 

In [29]:
data[data.model.isin(['ford f-250 sd', 'ford f250 super duty', 'ford f-250 super duty'])]['model'].value_counts()

model
ford f-250 super duty    1037
Name: count, dtype: int64

### 4. Hypothesis <a id='duplicates'></a>


In [30]:
data['condition'].value_counts()

condition
excellent    24773
good         20145
like new      4742
fair          1607
new            143
salvage        115
Name: count, dtype: int64

In [31]:
data['model'].value_counts().head(10)

model
ford f-150                    3326
chevrolet silverado 1500      2171
ram 1500                      1750
chevrolet silverado           1271
jeep wrangler                 1119
ram 2500                      1091
ford f-250 super duty         1037
toyota camry                   990
honda accord                   964
chevrolet silverado 2500hd     915
Name: count, dtype: int64

In [32]:
data['model_year'].value_counts().head(10)

model_year
2011.0    4017
2013.0    3994
2012.0    3982
2014.0    3540
2008.0    3462
2015.0    3403
2010.0    3168
2016.0    2993
2007.0    2854
2017.0    2437
Name: count, dtype: int64

In [33]:
data['fuel'].value_counts()                

fuel
gas         47288
diesel       3714
hybrid        409
other         108
electric        6
Name: count, dtype: int64

In [34]:
data['price'].value_counts().head(10)

price
1       798
6995    719
5995    655
4995    624
3500    620
7995    616
5500    615
4500    608
8995    549
6500    528
Name: count, dtype: int64

In [35]:
data['price'].describe()

count     51525.000000
mean      12132.464920
std       10040.803015
min           1.000000
25%        5000.000000
50%        9000.000000
75%       16839.000000
max      375000.000000
Name: price, dtype: float64

The price range starts from 1 to 375 000

In [36]:
data['cylinders'].value_counts()

cylinders
8.0     17712
6.0     17435
4.0     15499
10.0      549
5.0       294
3.0        34
12.0        2
Name: count, dtype: int64

In [37]:
data['odometer'].value_counts().head(10)

odometer
120000.0    579
121928.0    424
130000.0    357
107871.0    280
106000.0    254
125000.0    219
140000.0    216
150000.0    208
117000.0    201
131000.0    200
Name: count, dtype: int64

In [38]:
data['transmission'].value_counts()

transmission
automatic    46902
manual        2829
other         1794
Name: count, dtype: int64

In [39]:
data['type'].value_counts()

type
SUV            12405
truck          12353
sedan          12154
pickup          6988
coupe           2303
wagon           1541
mini-van        1161
hatchback       1047
van              633
convertible      446
other            256
offroad          214
bus               24
Name: count, dtype: int64

In [40]:
data['paint_color'].value_counts()

paint_color
white      10029
unknown     9267
black       7692
silver      6244
grey        5037
blue        4475
red         4421
green       1396
brown       1223
custom      1153
yellow       255
orange       231
purple       102
Name: count, dtype: int64

In [41]:
data['is_4wd'].value_counts()

is_4wd
0      25953
1.0    25572
Name: count, dtype: int64

In [42]:
data['date_posted'].value_counts().head(10)

date_posted
2019-03-17    186
2019-01-05    184
2018-09-20    178
2019-03-11    176
2018-10-09    174
2019-04-10    173
2018-08-05    173
2019-03-28    172
2018-05-23    171
2018-07-29    169
Name: count, dtype: int64

In [43]:
data['days_listed'].value_counts()

days_listed
18     959
24     950
22     945
19     941
20     934
      ... 
240      1
209      1
188      1
192      1
186      1
Name: count, Length: 227, dtype: int64